src/editors/EditorPageProvider.jsx

```diff
@@ -0,0 +1,88 @@
+import React, { useState, useEffect, useRef } from 'react';
+import PropTypes from 'prop-types';
+import { fetchBlockById, fetchUnitById, saveBlock } from './data/api';
+import EditorPageContext from './EditorPageContext';
+import { ActionStates } from './data/constants';
+
+/* This Component serves as a container for state for V2 editors,
+to avoid prop drilling for: saving, loading, and navigating away from content. */
+
+const EditorPageProvider = ({
+  blockType, courseId, blockId, studioEndpointUrl, children,
+}) => {
+  const editorRef = useRef(null);
+  const [blockValue, setBlockValue] = useState(null); // this is the intial block, as called in from the api.
+  const [blockError, setBlockError] = useState(null);
+  const [blockLoading, setBlockLoading] = useState(ActionStates.NOT_BEGUN);
+  const [unitUrl, setUnitUrlValue] = useState(null);
+  const [unitUrlError, setUnitUrlError] = useState(null);
+  const [unitUrlLoading, setUnitUrlLoading] = useState(ActionStates.NOT_BEGUN);
+  const [blockContent, setBlockContent] = useState(null); // This is the updated content to be saved via api call
+  const [saveResponse, setSaveResponse] = useState(null);
+  const [saveUnderway, setSaveUnderway] = useState(ActionStates.NOT_BEGUN);
+  useEffect(() => {
+    if (unitUrlLoading === ActionStates.NOT_BEGUN) {
+      fetchUnitById({ // pass as object for testing purposes.
+        setValue: setUnitUrlValue,
+        setError: setUnitUrlError,
+        setLoading: setUnitUrlLoading,
+      }, blockId, studioEndpointUrl);
+    }
+    if (blockLoading === ActionStates.NOT_BEGUN) {
+      fetchBlockById(
+        {
+          setValue: setBlockValue,
+          setError: setBlockError,
+          setLoading: setBlockLoading,
+        }, blockId, studioEndpointUrl,
+      );
+    }
+    if (saveUnderway === ActionStates.IN_PROGRESS) {
+      saveBlock(
+        blockId,
+        blockType,
+        courseId,
+        studioEndpointUrl,
+        blockContent,
+        { setInProgress: setSaveUnderway, setResponse: setSaveResponse },
+      );
+    }
+  }, [saveUnderway]);
+
+  return (
+    <EditorPageContext.Provider
```

Comment:
So I'd be careful here.  This provider will cause everything subscribing to it in its `children` to re-render on every single re-render here.  It's creating a new Object on every re-render, which trips all the `useContext` hooks below it in the tree, even if the data they depend on in the context hasn't changed.  Worst case scenario, this can cause a massive increase in re-rendering and slow your page down.

Context is best for data that changes infrequently, and when you're passing a lot of data into it like this, it may be best to memoize creation of the `value` so that you're not creating a new context value object each time this provider re-renders, if that makes sense.

Reply:
I may not have explained well here - I'm happy to do a better job synchronously if it's helpful.

Reply:
No that makes sense-- Its not really feasible to say don't use `useEffect` hooks in your editors, so I'll plan to use the `useMemo` hook around the value, and do some testing with logging as to how often things rerender. It seems to be effective at solving that problem, as we can make the dependancies of the memo hook just the loading states

Reply:
Alright so using memoization, I have `EditorFooter` (the consumer component which uses an effect hook) rendering on only 50% of renders from editorprovider, and the effect hooks are called much less often. Thanks!